ECOM-5281 Add retires for email sending while registering.

attiyaIshaque opened the pull request:
## [ECOM-5281](https://openedx.atlassian.net/browse/ECOM-5281)

### Description

Add Retires for email sending.

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the 👍.
- [x] @adampalay 
- [x] @awaisdar001 
- [x] @Qubad786

adampalay commented on common/djangoapps/student/views.py:
```diff
@@ -1793,21 +1795,33 @@ def create_account_with_params(request, params):
-                dest_addr,
-                exc_info=True
-            )
+        while True:
```
`while True`s always scare me :). Can we do something like:
```python
retries = settings.FEATURES.get('RETRY_MAX_ATTEMPTS')
while retries:
    try:
        ...
    except
        retries -= 1
```

attiyaIshaque replied:
@adampalay okay i will change it. :)

adampalay commented on common/djangoapps/student/views.py:
```diff
@@ -1793,21 +1795,33 @@ def create_account_with_params(request, params):
+                break
+            except Exception:  # pylint: disable=broad-except
+                retries += 1
+                if retries > settings.FEATURES.get('RETRY_MAX_ATTEMPTS'):
```
what is this value?

attiyaIshaque replied:
its RETRY_MAX_ATTEMPTS = 5

adampalay commented on common/djangoapps/student/views.py:
```diff
@@ -1793,21 +1795,33 @@ def create_account_with_params(request, params):
+                    exc_info=True
+                )
+            if retries > 1:
+                time.sleep(settings.RETRY_TIMEOUT)
```
this is always dangerous to do in process (and not in a background task). What's the retry timeout right now?

Would this make more sense as a background task? Like abstract the try/except block into a function `send_user_email` or something and turn it into a celery task?

jibsheet replied:
I would love to see mail in a background task, but not sure if it's too large a change for this fix.  I would loop in @jmbowman who's been designing some new celery work to support an eventual move for things like this.

attiyaIshaque replied:
Yeah it would be a celery task but as @jibsheet said there is large change for this.
But now I think we should remove this.

adampalay replied:
naive question: why shouldn't this be as straightforward as writing a function and decorating it with `@task` like we do here: https://github.com/edx/edx-platform/blob/bb36ab013f976b4e5b3dad12171f730dd22fc434/openedx/core/djangoapps/ccxcon/tasks.py#L22-L23

jibsheet replied:
@adampalay I think I wasn't sure if we should try to get all the emails out, which is a larger change.  From the ECOM ticket it sounds like even some password resets and other activation mails are failing to go out.

Do you want to look at splunk together and see if we have a sense for how often this is actually erroring out and where in the code?  We should be able to look at just the boto error to get a sense.

adampalay replied:
@jibsheet sure, let's grab 20 mins this afternoon?

adampalay commented on lms/envs/common.py:
```diff
@@ -362,6 +362,8 @@
     # a course to use saved grades.
     'ENABLE_SUBSECTION_GRADES_SAVED': False,
+
+    'RETRY_MAX_ATTEMPTS': 5,
```
I don't think this would belong in the `FEATURES` part of the settings—these are usually booleans to tell if a certain feature should be activated or not. Move this down with `RETRY_TIMEOUT`.

Also, let's make these names a little more descriptive:
`RETRY_ACTIVATION_EMAIL_MAX_ATTEMPTS`
`RETRY_ACTIVATION_EMAIL_TIMEOUT`

attiyaIshaque replied:
Okay @adampalay. I will change it :)

attiyaIshaque commented:
jenkins run python

attiyaIshaque commented:
@adampalay I have addressed the comments.

adampalay commented:
@attiyaIshaque , I thought the plan was to do this as a background task?

attiyaIshaque commented:
@adampalay okay. I think we should add a story for the background task.

attiyaIshaque commented:
@adampalay I have added celery task.

adampalay commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,61 @@
+    )
+    retries = settings.RETRY_ACTIVATION_EMAIL_MAX_ATTEMPTS
+    dest_addr = user.email
+    while retries:
```
this might be cleaner as a for loop

attiyaIshaque replied:
I feel like `while` is good fit in this particular case.

adampalay replied:
I know, you're right. I'm just always scared my while loops will result in an infinite loop 😨

adampalay commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,61 @@
+
+from edxmako.shortcuts import render_to_string
+
+log = logging.getLogger("edx.student")
```
why not `__name__`

adampalay commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,61 @@
+            ))
+            break
+        except Exception:  # pylint: disable=broad-except
+            retries -= 1
```
why no delay?

adampalay commented on common/djangoapps/student/views.py:
```diff
@@ -1775,37 +1775,8 @@ def create_account_with_params(request, params):
-                dest_addr,
-                exc_info=True
-            )
+        from student.tasks import send_activation_email
```
shouldn't this be imported at the top of the file?

attiyaIshaque commented:
@awaisdar001, @Qubad786 Please review. (Reviewers generated using random script)

attiyaIshaque commented:
@awaisdar001 , @Qubad786 ping

awaisdar001 commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,50 @@
+                    dest_addr,
+                    exc_info=True
+                )
+                raise
```
We were not raising the error previously, is this intentional ?

awaisdar001 commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,50 @@
+                )
+                raise
+            else:
+                log.info("Retrying sending email, attempt # {attempt} of {max_attempts} of MAX". format(
```
shouldn't we need to log the information of the destination addr here?

awaisdar001 commented on common/djangoapps/student/tests/test_tasks.py:
```diff
@@ -0,0 +1,48 @@
+    @mock.patch('student.tasks.log')
+    @mock.patch('django.contrib.auth.models.User')
+    def test_send_email(self, mock_user, mock_log):
+
```
no space please :)

awaisdar001 commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,50 @@
+@task()
+def send_activation_email(user, subject, message, from_address):
+    """
+        Sending an activation email to the users.
```
Unwanted spaces before `Sending ....`

awaisdar001 commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,50 @@
+    """
+        Sending an activation email to the users.
+    """
+    retries = settings.RETRY_ACTIVATION_EMAIL_MAX_ATTEMPTS
```
Just for a suggestion, wouldn't it be more readable if we just compare if our retries didn't acceded the maximum retries? 

In other words what I am saying rather than logging 
`"Retrying sending email, attempt # 5..."`
`"Retrying sending email, attempt # 4..."`

Make it such

```
max_retires = settings.RETRY_ACTIVATION_EMAIL_MAX_ATTEMPTS
retries = 0

while retries < max_retires:
    ...
    ...
    except Exception: 
        retries +=1
        if retries >= max_retries:
             log.info("Activataion Email has been sent to User ....")
             ....
       else:
           log.info("Retrying sending email, attempt #")  #now this would log correct attempts, 1, 2, 3, ... 


```
Well its just a suggestion, its upto you, :-)

awaisdar001 commented on common/djangoapps/student/tests/test_tasks.py:
```diff
@@ -0,0 +1,48 @@
+        from_address = "task_testing@edX.com"
+        mock_user.email_user.side_effect = Exception
+        email_max_attempts = settings.RETRY_ACTIVATION_EMAIL_MAX_ATTEMPTS
+        subject = "Task_test"
```
Doesn't seems like we need `subject` and `message` as variables as its only being used a 1 place. 
Also we should use `''` rather than `""`

awaisdar001 commented on common/djangoapps/student/tests/test_tasks.py:
```diff
@@ -0,0 +1,48 @@
+            send_activation_email(mock_user, subject, message, from_address)
+
+        # Asserts sending email retry logging.
+        for attempt in xrange(email_max_attempts - 1, 0, -1):
```
this `xrange` can be much simpler if we try out the suggestion.

attiyaIshaque commented:
jenkins run all

attiyaIshaque commented:
jenkins run bokchoy

attiyaIshaque commented:
@awaisdar001 I have addressed all the comments.

Qubad786 commented:
@attiyaIshaque , just came across http://celery.readthedocs.io/en/latest/userguide/tasks.html#retrying, we should use celery's  retry rather than implementing our own retry process. This is gonna make the code a lot simpler.

attiyaIshaque commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,52 @@
+                    exc_info=True
+                )
+                raise Exception
+            else:
```
Okay. I will change it.

attiyaIshaque commented:
jenkins run js

attiyaIshaque commented:
@awaisdar001, @Qubad786 Please review again.

awaisdar001 commented on common/djangoapps/student/tests/test_tasks.py:
```diff
@@ -0,0 +1,44 @@
+    @mock.patch('django.contrib.auth.models.User')
+    def test_send_email(self, mock_user, mock_log):
+        """
+        Test Send activation email exception
```
Can we describe what we actually testing here? i.e.  we are testing retries logic works as defined.

awaisdar001 replied:
Just a small nit. once done and tests are passed, I am ok with it. 👍

attiyaIshaque replied:
okay I will change it.

attiyaIshaque commented:
@adampalay celery's retry is really good; we can use it later but in this case, it's good to implement our own retry process.

attiyaIshaque commented:
jenkins run bokchoy

adampalay commented:
@attiyaIshaque why?

attiyaIshaque commented:
@adampalay let me gather my findings and I will come back to you.

attiyaIshaque commented:
@adampalay before reading celery's documentation I was not sure that either things will work for our case or not because everything is happening in backend. But after reading doc thoroughly , I have implemented in this case like, to add `max_retries `and logging etc. So, I have added the celery's retry. Kindly review it again. Thanks

adampalay commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,48 @@
+        ))
+        try:
+            self.retry(countdown=settings.RETRY_ACTIVATION_EMAIL_TIMEOUT, max_retries=max_retries)
+        except MaxRetriesExceededError:
```
I was just thinking—and maybe @Qubad786 can weigh in on this too—is it the wisest thing to retry for _all_ exceptions? Can we retry only on those exceptions where we know for sure that the email wasn't sent? Otherwise, if an exception is thrown after the email is sent, we might end up sending a whole bunch of activation emails to the user.

Qubad786 replied:
@adampalay, the way I see it, there no point of any exception raise after the email is sent successfully. If the email has been send successfully, the control returns to the log message, which i do not think can break – only response parsing is after success. Do you know any case where exception may be raised after email success?

adampalay replied:
There are a ton of exceptions whose baseclass is `Exception`. I think just not knowing of one isn't a good enough answer. @attiyaIshaque , what was the original exception that we were seeing? Can we just retry after that one?

attiyaIshaque replied:
@adampalay we don't know which type of exception is occurring here and why email is not sent, that's why I just added the logging to figure out why it's happening and adds main `Exception`.

adampalay replied:
@attiyaIshaque ok, I think it's wiser to figure out what the exception is first before adding retries

attiyaIshaque replied:
@adampalay yeah, you are right. But when I am trying to create an account on my local environment, every time the email is sent. In addition to this, there are more than one exceptions that may be raised causing `send_email` to fail. 

If we handle only this exception `"NoAuthHandlerFound"` which is mentioned in this [comment](https://openedx.atlassian.net/browse/ECOM-5281?focusedCommentId=211754&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-211754). then no other exception other than this will be caught.

adampalay replied:
@attiyaIshaque good thinking. Let's retry after `NoAuthHandlerFound` errors, and for all other errors just do what we were doing before

attiyaIshaque replied:
@adampalay I have added it.

attiyaIshaque commented on common/djangoapps/student/tasks.py:
```diff
@@ -0,0 +1,56 @@
+        log.info("Activataion Email has been sent to User {user_email}".format(
+            user_email=dest_addr
+        ))
+        # raise NoAuthHandlerFound()
```
okay. It's removed :)

attiyaIshaque commented:
jenkins run a11y

attiyaIshaque commented:
jenkins run bokchoy

attiyaIshaque commented:
jenkins run bokchoy

Qubad786 commented:
👍
